### PR TITLE
Quick fix for Pmono (and PmonoArtic) turning everything off in the default group

### LINKED ISSUE #5024: Pdef with fading STILL dies on some streams that generate (internal) cleanup, e.g. PmonoArtic
## Environment

* SuperCollider version: develop ac271e7f0 (not 3.11.0!!)
* Operating system: Lubuntu 20

## Steps to reproduce

It turns out the bug reported in #4954 is not fully fixed by #4996 **in develop**.

Specifically, this sequence is ok on develop now (as is in a 3.11.0 with 4996 manually applied).

```
Pdef(\one).fadeTime_(2);
Pdef(\one, Pbind(\dur, 0.3, \degree, 1)).play
Pdef(\one, Pfset({}, Pbind(\dur, 0.2, \freq, 440), { "bye".postln })).play
```

but if you follow it with

```
Pdef(\one, PmonoArtic(\default, \dur, 0.2, \freq, Pwhite(1,8) * 100 , \legato, 1)).play
Pdef(\one, PmonoArtic(\default, \dur, 0.2, \freq, Pwhite(1,8) * 100, \legato, 0.2 )).play 
Pdef(\one, PmonoArtic(\default, \dur, 0.2, \freq, Pwhite(1,8) * 100 , \legato, 1)).play 
```

the last line is still bugged on `develop` with

```
FAILURE IN SERVER /n_set Node 1229 not found
```

even though no errors are given in 3.11.0. It looks like it will take more investigation to find out why.

## Expected vs. actual behavior

The last line in the last example should not stop playing.

## TLDR: why it happens 

The reason why develop behaves differently than releases even with #4996 applied when running the 2nd snippet above is this March commit https://github.com/supercollider/supercollider/commit/83e4607cd41264526113fc940cd0ab2a0b2f0b31 to develop, merged in May #4787 . The code introduced in PmonoStreams by the latter [causes](https://github.com/supercollider/supercollider/issues/5024#issuecomment-645639699) the 2nd code snippet above to emit a `type: off` Event with an `id: nil`, which in turn causes everything running on the server at that point (in the default group) to be released!

Somewhat simpler reproducer for 5024:

```
Pdef(\one).fadeTime_(2);
Pdef(\one, PmonoArtic(\default, \dur, 0.2, \freq, 500, \legato, 0.1 )).play 
Pdef(\one, Pmono(\default, \freq, 240))
```

And the simplest reproducer I found so far: only PfadeOut is enough, no need for the EmbedOnce (or Ppar) part of what Pdef does on xfades:

```
Pmono(\default).play;
PfadeOut(PmonoArtic(\default, \dur, Pn(0.4, 1)), 0.3).play;
```

the 2nd line of that can be just

```
PfadeOut(Pmono(\default, \dur, 0.4), 0.3).play;
```

Pmono (and PmonoArtic, by inheritance) do a two-part setup of their cleanup, which is not fully specified at the time it is "armed", meaning registered with `cleanup.addFunction`. Pmono (and PmonoArtic) expect that the `~updatePmono` callback *will* be called to "backfill" the `id`, but PfadeOut somehow breaks that assumption. That's exactly what's happening in the above--easily verified by adding some `postln` to the `~updatePmono` callback. (And firing an event of `type: off` with a `nil` id turns everything off in the default group on the server.)

### PR DESCRIPTION
**Pull request by eleses**
## Purpose and Motivation

Fixes #5024 

Commit 83e4607cd merged in PR #4787 exposed an underlying problem in Pmono's two-part cleanup setup, which can (now at least) fire a `type: off` event with no `id` set, turning everything off in the default group, e.g. as in:

```
Synth(\default);
PfadeOut(Pmono(\default, \dur, 0.4), 0.3).play;
```

This is an "emergency fix" that prevents Pmono's cleanup from firing that `off` event if `id` is still not set by the time the
cleanup is executed. (The mechanism is also inherited by PmonoArtic.)

A better solution would be to rewrite Pmono so that it doesn't register an incompletely specified cleanup at all, but that is somewhat involved given the number of variables it would need to capture in a closure.

Also this PR will very likely cause a merge conflict with the solution currently envisaged for #4806 which needs to change the previous line in that Pmono cleanup setup code (to conform to the new API for cleanups), so please merge this PR as soon as practically possible.

## Types of changes

- Bug fix

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] This PR is ready for review

### CODE REVIEW

File: SCClassLibrary/Common/Streams/PmonoStreams.sc [hunk 1]
```diff
@@ -69,7 +69,7 @@ PmonoStream : Stream {
 					~type = \monoNote;
 					~instrument = pattern.synthName;
 					cleanup.addFunction(event, currentCleanupFunc = { | flag |
-						if (flag) { (id: id, server: server, type: \off,
+						if (flag && id.notNil) { (id: id, server: server, type: \off,
```
Comment (jamshark70) [inline SCClassLibrary/Common/Streams/PmonoStreams.sc:72]: One minor suggestion: if it's written `flag and: { id.notNil }` then there are two benefits: 1/ save one method dispatch (`&&` redirects to `and` -- unfortunately `&&` is not optimized as it is in C) and 2/ the second condition is short-circuited when `flag` is false. I'm pretty sure the cleanup is almost always called with `flag` being true, so short-circuiting isn't a big deal, but saving one method lookup is better (because we don't know how many events per second users may be generating, and how often they will rearticulate).

Comment (eleses) [reply SCClassLibrary/Common/Streams/PmonoStreams.sc:72]: Flag is false only when you hit cmd+. (CmdPeriod executed), as far as I know. So in the vast majority of cases/calls there will be no short-circuit, but your point is taken.

**Comment by jamshark70** [comment]
Per https://github.com/supercollider/supercollider/issues/5024#issuecomment-646655746, there's a very high likelihood that the stopgap proposed here wouldn't be necessary. So I'd suggest to hold off on merging -- I'm guessing >90% chance of closing this one without a merge.

**Comment by eleses** [comment]
> Per [#5024 (comment)](https://github.com/supercollider/supercollider/issues/5024#issuecomment-646655746), there's a very high likelihood that the stopgap proposed here wouldn't be necessary. So I'd suggest to hold off on merging -- I'm guessing >90% chance of closing this one without a merge.

[Per my subsequent comment there, your proposed alternative breaks `stop` on Pmono.](https://github.com/supercollider/supercollider/issues/5024#issuecomment-646680074), alas.

**Comment by jamshark70** [comment]
FWIW, I just verified that the problem is not reproducible after merging #5030, with no other changes (*and* stopping a regular Pmono is not broken with #5030 in place).

I'll wait a bit for other feedback, but AFAICS this PR doesn't need to be merged.

**Comment by eleses** [comment]
>  unsurprising as 5030 seems to incorporate the change here

That was intentional because of (1) the potential edit conflict since they changed adjacent lines of code... and (2) I wanted to check that they work together.

However, 5030 might take substantially longer to review/approve/merge.

**Comment by dyfer** [comment]
@eleses @jamshark70 should we close this then?

**Comment by jamshark70** [comment]
Hm... "unsurprising as 5030 seems to incorporate the change here."

So if we keep it in 5030, then this PR would be redundant. However 5030 seems to include changes not related to CleanupThunk/CallOnce. It might be worth narrowing the scope.

I suspect we want to keep it -- a cleanup event with a nil ID is not good, regardless of how we got there.

**Comment by joshpar** [comment]
@jamshark70 and @dyfer - OK to close?

**Comment by jamshark70** [comment]
I confirmed that the issue is still reproducible in current develop, *but* it is not reproducible after applying #5386 (de-dup EventStreamCleanup). This was a fortunate accident -- I had applied 5386 locally to check the behavior of CallOnce().valueArray, and forgotten to switch back to develop before checking this case. It worked fine with 5386.

So, if we have a commitment to merge #5386, then this one can be closed.
